### ISSUE TITLE
xah-delete-backward-char-or-bracket-text get's very slow on large latex file

### ISSUE THREAD

**Issue by thorstengrothe** [xahlee/xah-fly-keys#19]
Hi,
the above command get's very slow after using it a couple of times . CPU-Load is about 40% and it seems that emacs freezes for some moments. I work on a large latex file with 6500 lines. With smaller files this does not happen. I got

```
(defun xah-delete-backward-char-or-bracket-text ()
 ...
Version 2017-03-13"

 
```
from xah-fly-keys. My emacs is 25.1.1 on Archlinux/Manjaro.

I found out, that it only happens with *.tex files. If I rename the file to *.txt the command works in normal speed, so it has something to do with auctex, but I don't know what to do any further. I disabled font-lock-mode and reftex-mode and autofill-mode in auctex but that does not help....

I attached the large file as *.txt because *.tex is not supported. 

Can anybody confirm this behaviour?

Thorsten
[VWL_TEST.txt](https://github.com/xahlee/xah-fly-keys/files/951628/VWL_TEST.txt)

**Comment by xahlee** [comment]
2 things i like to know:

• In the same file, if you have parenthesis within a line, say just 20
characters apart, does it still happen?

• in the same file, if you M-x font-lock-mode to disable the font lock,
does it still happen?

thanks.

**Comment by xahlee** [comment]
\begin{center}
...
\end{center}

and cursor is all the way at the end of right curly bracket, and
xah-delete-backward-char-or-bracket-text
will delete the whole bracket text at the \begin line.

is this what you experience too?

this happens because apparently LaTeX mode has text properties that
indicate matching begin/end or something. So, the normal behavior of
backward-sexp
forward-sexp
mark-sexp is not the same as lisp parens.

e.g. when cursor is somewhere in \begin, then M-x mark-sexp will select the
whole begin/end block.

am not sure what to do about it.

what do you think should be the proper behavior?

---------

for the slowness, can you give me a line number and char position to
duplicate that?

**Comment by thorstengrothe** [comment]
Xah,

first of all I disabled font-lock-mode and flyspell. To reproduce do the following:

1. Goto line 45-59 and try to delete, looks like this:
 ```
    BWL/VWL,
    Bedürfnisse,
    Güter,
    Ökonomisches Prinzip,
    Produktionsfaktoren,
    MKK,
    Geld,
    Warenkorb,
    Wirtschaftskreislauf,
    Markt,
    Elastizitäten,
    Marktformen,
    VGR,
    Investitionsarten,
    Magisches Viereck
```
2. Go to the end of the document **(\end{document})** and paste this:

`Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.`

Go to last character of the last line and try it again, here emacs get's unusable slow.

Is this the same for you?

Thorsten

**Comment by xahlee** [comment]
no slowness happens here. It's instantaneous.  I'm on LaTeX mode. So am guessing it's auctex mode.
Try in LaTeX mode and see if it still happens.

**Comment by thorstengrothe** [comment]
Strange, I disabled almost all other packages and tried LaTeX mode, it's the same, maybe it has something to do with my emacs config, but I have no clue where to look..

**Comment by thorstengrothe** [comment]
Tested with the large org.el file. In the middle of the document everything is fine.
When you go to the end of the file

`;;; org.el ends here`

it's the same. When I insert some text at any place in the document and execute the command it happens again. When I execute the command **before adding text** it works fine, very strange....

[org.el.txt](https://github.com/xahlee/xah-fly-keys/files/953131/org.el.txt)

**Comment by xahlee** [comment]
just to make this clear. Here's the steps i did.

* open VWL_TEST.txt
* go to line 60, place cursor at column 3
* press e (dvorak) to call xah-delete-backward-char-or-bracket-text
* the whole bracked and inner text are deleted. I'm left with the line “  use-topics = ▮”
* move cursor to end of file
* paste the “Lorem ...” paragraph.
* move cursor to line 6633, column 14, which is “\end{document}▮”
* press e (dvorak) to call xah-delete-backward-char-or-bracket-text.
* my cursor is now at line 48, showing “▮{document}”. (the “\begin” in front is deleted)

the whole thing is instantaneous.

**Comment by xahlee** [comment]
the org.el.txt you sent... 

it's important to note which major mode is on to test this.

i tried the org.el.txt in text-mode and org-mode, wasn't able to reproduce slowness. I don't  see how slowness problem could happen in text mode or org mode. Especially text mode.

Do you mean this happens in text mode too? Because if it does, this eliminates major mode issues. Then, it must be some package you have, i think.

**Comment by xahlee** [comment]
the command xah-delete-backward-char-or-bracket-text
basically just call the following to do its work:

<C-M-left> runs the command backward-sexp
<C-M-right> runs the command forward-sexp

so, you could just try backward-sexp in those situations, and see if that
is slow. If so, that probably means something in your init is not optimal.

otherwise, if backward-sexp works fine but
xah-delete-backward-char-or-bracket-text cause slowness, then the problem
is xah-delete-backward-char-or-bracket-text.

**Comment by thorstengrothe** [comment]
Oh sorry I forgot to explain that. I renamed org.el.txt because *.el files are not supported to upload. 

For the following tests I used the uploaded **org.el.txt** and renamed it back to **org.el!**

I tested the following commands from the end of the document: **line 25203: ;;; org.el ends here** and keep it pressed for a couple of seconds:

 _<C-M-left> backward-sexp
 <C-M-right> forward-sexp_

Both are running really fast and instantaneous, so no problem. 

Next I closed the file without saving and tested the same with _xah-delete-backward-char-or-bracket-text_ 

again from **line 25203: ;;; org.el ends here** and keep it again pressed for a couple of seconds: slowness comes back. 

Next I go to line 1, 18, 208 and tried _xah-delete-backward-char-or-bracket-text_ --> not slowness.

Next I go to line 25202 and entered this Text: 

> ABCABCBCABCABC

_forward/backward-sexp_ are running fast with _xah-delete-backward-char-or-bracket-text_ the slowness comes back. Is that clear enough explained to understand the problem?

**Comment by xahlee** [comment]
ok, do you mean holding the key all this while?

if you hold the key of
xah-delete-backward-char-or-bracket-text,
it'll be slow because each key press it's calling 
make-syntax-table
and
syntax-ppss
(i forgot to mention that).
After doing the above, then it possibly calls backward-sexp.
It calls syntax-ppss because we wanted to see if we are in comment or string, if so, we just want to delete a single char. syntax-ppss is relatively slow.
It calls make-syntax-table so that we want to be sure brackets such as ‹› are considered a bracke in any current major mode...

So, is the whole issue about holding a key down?

I'm kinda lost what's the issue.

The command xah-delete-backward-char-or-bracket-text isn't stable, as in, there are lots of room on what it should be or where it goes. As i mentioned earlier, the behavior of deleting unexpected part of text in LaTeX mode is an issue. But is that an issue you are concerned? Then, about the slowness, which i don't know what you mean. Then, 3rd point being the key held down issue, if that is the real issue we are discussion. 


if the issue is about holding down the key being slow, i'm thinking perhaps the key should just be normal delete backward char. While the key for delete forward char will be the delete bracket behavior...

**Comment by xahlee** [comment]
can you try this? see if you like this version.

(defun xah-delete-backward-char-or-bracket-text ()
  "Delete backward 1 character, but if it's a \"quote\" or bracket ()[]{}【】「」 etc, delete bracket and the inner text, push the deleted text to `kill-ring'.

What char is bracket is determined by current syntax table.

If `universal-argument' is called first, do not delete inner text.

URL `http://ergoemacs.org/emacs/emacs_delete_backward_char_or_bracket_text.html'
Version 2017-04-25 26437"
  (interactive)

  (if (and delete-selection-mode (region-active-p))
      (delete-region (region-beginning) (region-end))
    (cond
     ((looking-back "\\s)" 1)
      (progn
        (backward-sexp)
        (xah-delete-matching-brackets (not current-prefix-arg))))
     ((looking-back "\\s(" 1)
      (progn
        (backward-char )
        (xah-delete-matching-brackets (not current-prefix-arg))))
     ((looking-back "\\s\"" 1)
      (if (nth 3 (syntax-ppss))
          (progn
            (backward-char )
            (xah-delete-matching-brackets (not current-prefix-arg)))
        (progn
          (backward-sexp)
          (xah-delete-matching-brackets (not current-prefix-arg)))))
     (t (delete-char -1)))))

can just put this in emacs init, after the code of xah fly keys.

**Comment by thorstengrothe** [comment]
Yes, I guess the whole thing is because of holding down the **d** key (qwerty here). I tested your new function but unfortunately the problem is still there. But I can live with that, no problem. I agree fully with you to assign the command 

> delete-backward-char

to **d** and assign

> xah-delete-backward-char-or-bracket-text

to another key for brackets stuff. I did that, in my case to **G6** on my Logitech G710+ keyboard and it works fine. I guess it's easier to treat both commands different, one for (fast) deletion of text and one for brackets stuff with text inside.

For me the problem is solved!

Thanks for your awesome support!

Thorsten
